runtime/vm/xa64/UpcallThunkGen.cpp

```diff
@@ -35,32 +35,1219 @@ extern "C" {
+						MOVSS_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, xmm0)
+					} else {
+						MOVSD_TREG_mRSP_DISP32m(thunkCursor, xmm0, frameSize + preservedRegisterAreaSize + 8 + memParmCursor)
+						MOVSD_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, xmm0)
+					}
+
+					memParmCursor += STACK_SLOT_SIZE;
+				}
+
+				frameOffsetCursor += STACK_SLOT_SIZE;
+				break;
+			}
+			default:
+			{
+				// Handle structs passed in registers.  Structs passed in memory will be handled
+				// after all other parameters are processed to avoid the complication of preserving
+				// registers implicitly required for REP MOVSB.
+				X64StructPassingMechanism mechanism = analyzeStructParm(gprRegParmCount, fprRegParmCount, sigArray[i]);
+				switch (mechanism) {
+					case PASS_STRUCT_IN_MEMORY:
+					{
+						// Record the source and destination offsets and the number of bytes to copy
+						structParmInMemory[numStructsPassedInMemoryCursor].memParmCursor = memParmCursor;
+						structParmInMemory[numStructsPassedInMemoryCursor].frameOffsetCursor = frameOffsetCursor;
+						structParmInMemory[numStructsPassedInMemoryCursor].sizeofStruct = sigArray[i].sizeInByte;
+
+						I_32 roundedStructSize = ROUND_UP_TO_SLOT_MULTIPLE(sigArray[i].sizeInByte);
+						memParmCursor += roundedStructSize;
+						frameOffsetCursor += roundedStructSize;
+						numStructsPassedInMemoryCursor += 1;
+						break;
+					}
+					case PASS_STRUCT_IN_ONE_FPR:
+					{
+						MOVSD_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, fprParmRegs[fprRegParmCount])
+						fprRegParmCount += 1;
+						frameOffsetCursor += STACK_SLOT_SIZE;
+						break;
+					}
+					case PASS_STRUCT_IN_TWO_FPR:
+					{
+						MOVSD_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, fprParmRegs[fprRegParmCount])
+						fprRegParmCount += 1;
+						frameOffsetCursor += STACK_SLOT_SIZE;
+						MOVSD_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, fprParmRegs[fprRegParmCount])
+						fprRegParmCount += 1;
+						frameOffsetCursor += STACK_SLOT_SIZE;
+						break;
+					}
+					case PASS_STRUCT_IN_ONE_GPR_ONE_FPR:
+					{
+						S8_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, gprParmRegs[gprRegParmCount])
+						gprRegParmCount++;
+						frameOffsetCursor += STACK_SLOT_SIZE;
+						MOVSD_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, fprParmRegs[fprRegParmCount])
+						fprRegParmCount += 1;
+						frameOffsetCursor += STACK_SLOT_SIZE;
+						break;
+					}
+					case PASS_STRUCT_IN_ONE_FPR_ONE_GPR:
+					{
+						MOVSD_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, fprParmRegs[fprRegParmCount])
+						fprRegParmCount += 1;
+						frameOffsetCursor += STACK_SLOT_SIZE;
+						S8_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, gprParmRegs[gprRegParmCount])
+						gprRegParmCount++;
+						frameOffsetCursor += STACK_SLOT_SIZE;
+						break;
+					}
+					case PASS_STRUCT_IN_ONE_GPR:
+					{
+						S8_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, gprParmRegs[gprRegParmCount])
```

Comment:
similar comment as float/double:  what about the remaining GPR content is smaller than 8-byte?